test/widgets/action_sheet_test.dart

```diff
@@ -99,46 +107,244 @@ void main() {
+            emojiCode: '1f4a4',
+            emojiName: 'zzz'),
+        ]);
+
+        tester.widget(searchFieldFinder);
+        await tester.enterText(searchFieldFinder, 'z');
+        await tester.pump();
+
+        check(tester.widgetList<EmojiPickerListEntry>(find.byType(EmojiPickerListEntry))).deepEquals([
+          conditionEmojiListEntry(
+            emojiType: ReactionType.zulipExtraEmoji,
+            emojiCode: 'zulip',
+            emojiName: 'zulip'),
+          conditionEmojiListEntry(
+            emojiType: ReactionType.unicodeEmoji,
+            emojiCode: '1f4a4',
+            emojiName: 'zzz'),
+          conditionEmojiListEntry(
+            emojiType: ReactionType.realmEmoji,
+            emojiCode: '1',
+            emojiName: 'buzzing'),
+        ]);
+
+        tester.widget(searchFieldFinder);
+        await tester.enterText(searchFieldFinder, 'zz');
+        await tester.pump();
+
+        check(tester.widgetList<EmojiPickerListEntry>(find.byType(EmojiPickerListEntry))).deepEquals([
+          conditionEmojiListEntry(
+            emojiType: ReactionType.unicodeEmoji,
+            emojiCode: '1f4a4',
+            emojiName: 'zzz'),
+          conditionEmojiListEntry(
+            emojiType: ReactionType.realmEmoji,
+            emojiCode: '1',
+            emojiName: 'buzzing'),
+        ]);
+
+        debugNetworkImageHttpClientProvider = null;
       });
-      await tapButton(tester);
-      await tester.pump(Duration.zero); // error arrives; error dialog shows
 
-      await tester.tap(find.byWidget(checkErrorDialog(tester,
-        expectedTitle: 'Adding reaction failed',
-        expectedMessage: 'Invalid message(s)')));
+      testWidgets('adding success', (tester) async {
+        final message = eg.streamMessage();
+        await setupEmojiPicker(tester, message: message, narrow: TopicNarrow.ofMessage(message));
+
+        connection.prepare(json: {});
+        await tester.tap(find.text('\u{1f4a4}')); // 'zzz' emoji
+        await tester.pump(Duration.zero);
+
+        check(connection.lastRequest).isA<http.Request>()
+          ..method.equals('POST')
+          ..url.path.equals('/api/v1/messages/${message.id}/reactions')
+          ..bodyFields.deepEquals({
+              'reaction_type': 'unicode_emoji',
+              'emoji_code': '1f4a4',
+              'emoji_name': 'zzz',
+            });
+
+        debugNetworkImageHttpClientProvider = null;
+      });
+
+      testWidgets('request has an error', (tester) async {
+        final message = eg.streamMessage();
+        await setupEmojiPicker(tester, message: message, narrow: TopicNarrow.ofMessage(message));
+
+        connection.prepare(httpStatus: 400, json: {
+          'code': 'BAD_REQUEST',
+          'msg': 'Invalid message(s)',
+          'result': 'error',
+        });
+        await tester.tap(find.text('\u{1f4a4}')); // 'zzz' emoji
+        await tester.pump(Duration.zero); // error arrives; error dialog shows
```

Comment:
Let's adapt this test so that it would successfully catch if we accidentally broke in the future the careful work you did to properly pass down `pageContext`: https://github.com/zulip/zulip-flutter/pull/1103#discussion_r1877511231

(As Hixie likes to say: the purpose of a test is to make sure we don't accidentally revert your changes.)

As is, if you apply this diff, no tests fail:
```diff
           itemBuilder: (context, i) => EmojiPickerListEntry(
-            pageContext: widget.pageContext,
+            pageContext: context,
```

You can make the test catch that by doing this:
```suggestion
        await tester.pump(); // register tap
        await tester.pump(Duration(seconds: 1)); // emoji picker animates away
        await tester.pump(Duration(seconds: 1)); // error arrives; error dialog shows
```

and adding `delay: Duration(seconds: 2)` to the `prepare` call.